### PR TITLE
Create initialization callback hook

### PR DESCRIPTION
**Pull request by cgmckeever**
# Objective
As [pondered](https://github.com/nrwiersma/ConfigManager/pull/89#discussion_r468972398) create an initialization callback hook when `magicBytes` has not been set, therefore indicating a cold start.

### CODE REVIEW

File: src/ConfigManager.cpp [hunk 1]
```diff
@@ -23,32 +23,41 @@ void ConfigManager::setup() {
...
 
-      startApi();
+        startApi();
+      }
+    } else {
+      DebugPrintln(F("No SSID found"));
     }
   } else {
     // We are at a cold start, don't bother timing out.
+    if (initCallback) {
```
Comment (cgmckeever) [inline src/ConfigManager.cpp:58]: When the magic is not set, the `initCallback` is called which can do things .. current use case is set some config defaults

File: src/ConfigManager.cpp [hunk 2]
```diff
@@ -235,7 +248,7 @@ void ConfigManager::clearWifiSettings(bool reboot) {
   memset(password, 0, PASSWORD_LENGTH);
 
   DebugPrintln(F("Clearing WiFi connection."));
-  storeWifiSettings(ssid, password, true);
+  storeWifiSettings(ssid, password, false);
```
Comment (cgmckeever) [inline src/ConfigManager.cpp:251]: I didnt see the reason that this was begin flipped. Maybe it was a remnant of some code changes ? Now Im thinking this is more of the flag that the memory has been initialized?

File: src/ConfigManager.cpp [hunk 3]
```diff
@@ -420,6 +440,8 @@ void ConfigManager::streamFile(const char* file, const char mime[]) {
 }
 
 void ConfigManager::handleAPGet() {
+  DebugPrint(F("Index Page: "));
+  DebugPrintln(F(apFilename));
```
Comment (nrwiersma) [inline src/ConfigManager.cpp:444]: You cannot `F()` a variable like this.

File: src/ConfigManager.cpp [hunk 4]
```diff
@@ -410,7 +429,8 @@ void ConfigManager::streamFile(const char* file, const char mime[]) {
 
   File f = SPIFFS.open(file, "r");
   if (!f) {
-    DebugPrintln(F("file open failed"));
+    DebugPrint(F("file open failed "));
+    DebugPrintln(F(file));
```
Comment (nrwiersma) [inline src/ConfigManager.cpp:433]: Here too.

File: src/ConfigManager.cpp (hunk 3 above)
Comment (cgmckeever) [reply src/ConfigManager.cpp]: those were 💩

**Comment by nrwiersma** [comment]
The code looks good. Could you perhaps include the `initCallback` in the `simple` example to show its use? Perhaps to init the defaults.

File: examples/simple/simple.ino [hunk 5]
```diff
@@ -35,9 +40,14 @@ void setup() {
     configManager.addParameter("password", config.password, 20, set);
     configManager.addParameter("version", &meta.version, get);
 
+    // Create custom routes to serve via callback hooks
     configManager.setAPCallback(createCustomRoute);
     configManager.setAPICallback(createCustomRoute);
 
+    // Create a initialization hook
+    // Will run until a memory write is performed.
+    configManager.setInitCallback(initCallback);
```
Comment (cgmckeever) [inline examples/simple/simple.ino:49]: @nrwiersma

File: src/ConfigManager.cpp [hunk 6]
```diff
@@ -325,13 +343,18 @@ JsonObject ConfigManager::asJson() {
   return obj;
 }
 
+bool ConfigManager::commitChanges() {
+  EEPROM.put(0, magicBytes);
+  return EEPROM.commit();
+}
```
Comment (cgmckeever) [inline src/ConfigManager.cpp:349]: @nrwiersma This is newly added. Both storeWifi and writeConfig use this hook to do the final commit, which then ensures the magicByte is set to indicate initialized (prior this was only set on storing wifi)

**Comment by nrwiersma** [comment]
Thanks, great feature addition.

**Pull request merged**